tao_compiler/mlir/disc/transforms/quantized_dot_rewriter.cc

```diff
@@ -0,0 +1,209 @@
+#include "tensorflow/compiler/mlir/disc/transforms/PassDetail.h"
+#include "tensorflow/compiler/mlir/disc/transforms/placement_utils.h"
+
+namespace mlir {
+namespace disc_ral {
+
+namespace {
+
+// It does not support transpose of batching dimension and other dimensions
+// together now.
+// TODO: it may support the transpose between batching dimensions only, or
+// transpose of all batching dimensions together with the minor dimension.
+static inline bool isNonBatchingTransposeTensorValue(
+    Value val, SmallVector<int64_t, 4>& permutation,
+    std::unordered_set<int64_t> batching_dims) {
+  if (not val.getDefiningOp()) {
+    return false;
+  }
+  permutation.clear();
+  if (auto transpose = dyn_cast<mhlo::TransposeOp>(val.getDefiningOp())) {
+    for (auto& en :
+         llvm::enumerate(transpose.permutation().getValues<int64_t>())) {
+      if (en.index() != en.value()) {
+        if (batching_dims.find(en.index()) != batching_dims.end()) {
+          return false;
+        }
+      }
+      permutation.push_back(en.value());
+    }
+  }
+  return !permutation.empty();
+}
+
+RankedTensorType GetTransposeOutputType(
+    Value value, const SmallVectorImpl<int64_t>& transpose_permutation,
+    OpBuilder& b) {
+  // Compute the resulting shape.
+  llvm::SmallVector<int64_t, 4> transposed_shape;
+  ShapedType input_type = value.getType().cast<ShapedType>();
+  auto input_shape = input_type.getShape();
+  for (int64_t val : transpose_permutation) {
+    transposed_shape.push_back(input_shape[val]);
+  }
+  return RankedTensorType::get(transposed_shape, input_type.getElementType());
+}
+
+Operation* InsertTranspose(
+    mlir::mhlo_disc::QuantizedDotGeneralOp op, Value value,
+    const SmallVectorImpl<int64_t>& transpose_permutation, OpBuilder& b) {
+  auto transpose_permutation_attr =
+      GetI64ElementsAttr(transpose_permutation, &b);
+
+  auto transpose_type = GetTransposeOutputType(value, transpose_permutation, b);
+  auto transpose_op = b.create<mhlo::TransposeOp>(
+      op.getLoc(), transpose_type, value, transpose_permutation_attr);
+
+  if (auto attr = op->getAttr(placement_utils::kDiscPlaceAssignment))
+    transpose_op->setAttr(placement_utils::kDiscPlaceAssignment, attr);
+
+  return transpose_op;
+}
+
+// The basic logic of this pass is to insert transpose op to ensure the qgemm op
+// having expected format.
+struct QuantDotTransposeConvert
+    : public OpRewritePattern<mlir::mhlo_disc::QuantizedDotGeneralOp> {
+  explicit QuantDotTransposeConvert(MLIRContext* context)
+      : OpRewritePattern(context) {}
+
+  LogicalResult matchAndRewrite(mlir::mhlo_disc::QuantizedDotGeneralOp op,
+                                PatternRewriter& rewriter) const override {
+    bool onGpu = placement_utils::isGpuMhlo(op);
+    // Only need to transpose weight on gpu currently
+    if (onGpu == false) {
+      return failure();
+    }
+
+    if (op.weight().getDefiningOp<mhlo::TransposeOp>()) {
```

Comment:
The candidate `QuantizedDotGeneralOp` ops are those having `k*n` layout, not those having `TransposeOp` operands. It's not reliable to only check the existence of the transpose op (e.g. transpose + dot op having n x k layout should also be taken into consideration).

Reply:
Make sense, done.